jit-service/src/main/java/org/dcsa/jit/service/TransportCallService.java

```diff
@@ -0,0 +1,122 @@
+package org.dcsa.jit.service;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.dcsa.jit.mapping.EnumMappers;
+import org.dcsa.jit.mapping.LocationMapper;
+import org.dcsa.jit.persistence.entity.TransportCall;
+import org.dcsa.jit.persistence.entity.Vessel;
+import org.dcsa.jit.persistence.entity.Voyage;
+import org.dcsa.jit.persistence.repository.FacilityRepository;
+import org.dcsa.jit.persistence.repository.LocationRepository;
+import org.dcsa.jit.persistence.repository.ServiceRepository;
+import org.dcsa.jit.persistence.repository.TransportCallRepository;
+import org.dcsa.jit.transferobjects.TimestampTO;
+import org.dcsa.skernel.domain.persistence.entity.Facility;
+import org.dcsa.skernel.domain.persistence.entity.Location;
+import org.dcsa.skernel.errors.exceptions.ConcreteRequestErrorMessageException;
+import org.springframework.stereotype.Service;
+
+import javax.transaction.Transactional;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class TransportCallService {
+  private final TransportCallRepository repository;
+  private final LocationRepository locationRepository;
+  private final FacilityRepository facilityRepository;
+  private final ServiceRepository serviceRepository;
+  private final EnumMappers enumMappers;
+  private final LocationMapper locationMapper;
+
+  @Transactional
+  public TransportCall ensureExists(TimestampTO timestampTO) {
+    validateTimestampTO(timestampTO);
+    return findTransportCall(timestampTO)
+      .orElseGet(() -> createTransportCall(timestampTO));
+  }
+
+  private void validateTimestampTO(TimestampTO timestampTO) {
+    if (timestampTO.modeOfTransport() == null) {
+      throw ConcreteRequestErrorMessageException.invalidInput("modeOfTransport must be given");
+    }
+    if (timestampTO.exportVoyageNumber() != null ^ timestampTO.importVoyageNumber() != null) {
+      throw ConcreteRequestErrorMessageException.invalidInput("exportVoyageNumber and importVoyageNumber must be given together or not at all");
+    }
+    if (timestampTO.carrierServiceCode() == null) {
+      throw ConcreteRequestErrorMessageException.invalidInput("Cannot create timestamp where service code is missing");
+    }
+    if (timestampTO.exportVoyageNumber() == null) {
+      throw ConcreteRequestErrorMessageException.invalidInput("Cannot create timestamp where voyage number (carrierVoyageNumber OR exportVoyageNumber + importVoyageNumber) is missing");
+    }
```

Comment:
I believe you can have timestamps where these fields are null. (Obviously not optimal, but the standard allows it). Related `findTransportCall` has a case where the TimestampTO is checked for having a null service code (which only used to give a more precise error).